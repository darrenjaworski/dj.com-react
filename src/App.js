import React from "react";
import {
  BrowserRouter as Router,
  Route,
  NavLink,
  Switch
} from "react-router-dom";

import styled from "@emotion/styled";
import { ThemeProvider } from "emotion-theming";
import themes from "./theme/theme";
import { Sun, Moon } from "react-feather";

import Home from "./routes/Home";
import Projects from "./routes/Projects";
import Journalism from "./routes/Journalism";
import FourOhFour from "./routes/404";

const AppWrapper = styled.div`
  background: ${props => props.theme.bgColor};
  transition: background 0.3s;
  min-height: 100vh;
`;

const Container = styled.div`
  max-width: 45rem;
  margin: auto;
  padding: 1rem;
`;

const Nav = styled.nav`
  padding: 1rem 0;
`;

const Toggle = styled.button`
  position: relative;
  background: transparent;
  color: ${props => props.theme.color};
  transition: all 0.5s;
  border: none;
  margin-top: 1rem;
`;

function getRandomIntInclusive(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min; //The maximum is inclusive and the minimum is inclusive
}

class App extends React.Component {
  constructor() {
    super();
    this.state = {
      theme: "dark",
      image: undefined
    };
  }

  componentDidMount() {
    const themeChoice = localStorage.getItem("theme");
    if (themeChoice) {
      this.setState({ theme: themeChoice });
    }
    const width = getRandomIntInclusive(500, 720);
    const height = getRandomIntInclusive(300, 600);
    const url = `https://www.fillmurray.com/g/${width}/${height}`;

    const image = new Image();
    image.src = url;
    this.setState({ image: url });
  }

  toggleTheme = () => {
    if (this.state.theme === "dark") {
      this.setState({ theme: "light" });
      localStorage.setItem("theme", "light");
    } else {
      this.setState({ theme: "dark" });
      localStorage.setItem("theme", "dark");
    }
  };

  render() {
    const { theme, image } = this.state;
    const toggleTheme = this.toggleTheme;
    const NavStyle = {
      color: `${themes[theme].linkColor}`,
      transition: "color 0.5s",
      textDecoration: "underline",
      marginRight: "0.5rem",
      fontSize: "1.25rem"
    };

    return (
      <Router>
        <ThemeProvider theme={themes[theme]}>
          <AppWrapper>
            <Container>
              <header>
                <Nav>
                  <NavLink style={NavStyle} to="/">
                    home
                  </NavLink>
                  <NavLink style={NavStyle} to="/projects">
                    projects
                  </NavLink>
                  <NavLink style={NavStyle} to="/journalism">
                    journalism
                  </NavLink>
                </Nav>
              </header>
              <main>
                <Switch>
                  <Route exact path="/" component={Home} />
                  <Route path="/projects" component={Projects} />
                  <Route path="/journalism" component={Journalism} />
                  <Route component={() => <FourOhFour image={image} />} />
                </Switch>
              </main>
              <Toggle className="toggle" onClick={toggleTheme}>
                {theme === "dark" ? <Sun /> : <Moon />}
              </Toggle>
            </Container>
          </AppWrapper>
        </ThemeProvider>
      </Router>
    );
  }
}

export default App;
