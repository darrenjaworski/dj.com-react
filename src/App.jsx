import React, { useState, useEffect } from "react";
import {
  BrowserRouter as Router,
  Route,
  NavLink,
  Switch
} from "react-router-dom";

import styled from "@emotion/styled";
import { ThemeProvider } from "emotion-theming";
import { Sun, Moon } from "react-feather";

import themes from "./theme/theme";
import Home from "./routes/Home";
import Projects from "./routes/Projects";
import Journalism from "./routes/Journalism";
import FourOhFour from "./routes/404";
import GeneticFrogs from "./routes/genetic-frogs/GeneticFrogs";

const AppWrapper = styled.div`
  background: ${(props) => props.theme.bgColor};
  transition: background 0.3s;
  min-height: 100vh;
`;

const Container = styled.div`
  max-width: 45rem;
  margin: auto;
  padding: 1rem;
`;

const Nav = styled.nav`
  padding: 1rem 0;
`;

const Toggle = styled.button`
  position: relative;
  background: transparent;
  color: ${(props) => props.theme.color};
  transition: all 0.5s;
  border: none;
  margin-top: 1rem;
`;

function getRandomIntInclusive(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  // The maximum is inclusive and the minimum is inclusive
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getRandomBillMurray() {
  const width = getRandomIntInclusive(500, 720);
  const height = getRandomIntInclusive(300, 600);
  const url = `https://www.fillmurray.com/g/${width}/${height}`;

  const image = new Image();
  image.src = url;

  return url;
}

const App = () => {
  const [theme, setTheme] = useState(
    localStorage.getItem("theme") ? localStorage.getItem("theme") : "dark"
  );
  const [image] = useState(getRandomBillMurray());

  const toggleTheme = () =>
    theme === "dark" ? setTheme("light") : setTheme("dark");

  useEffect(() => localStorage.setItem("theme", theme));

  const NavStyle = {
    color: `${themes[theme].linkColor}`,
    transition: "color 0.5s",
    textDecoration: "underline",
    marginRight: "0.5rem",
    fontSize: "1.25rem",
  };

  return (
    <Router>
      <ThemeProvider theme={themes[theme]}>
        <AppWrapper data-testid="body-theme">
          <Container>
            <header>
              <Nav>
                <NavLink style={NavStyle} to="/">
                  home
                </NavLink>
                <NavLink style={NavStyle} to="/projects">
                  projects
                </NavLink>
                <NavLink style={NavStyle} to="/journalism">
                  journalism
                </NavLink>
              </Nav>
            </header>
            <main>
              <Switch>
                <Route exact path="/" component={Home} />
                <Route path="/projects" component={Projects} />
                <Route path="/journalism" component={Journalism} />
                <Route path="/frogs" component={GeneticFrogs} />
                <Route component={() => <FourOhFour image={image} />} />
              </Switch>
            </main>
            <Toggle data-testid="toggle" onClick={toggleTheme}>
              {theme === "dark" ? <Sun /> : <Moon />}
            </Toggle>
          </Container>
        </AppWrapper>
      </ThemeProvider>
    </Router>
  );
};

export default App;
