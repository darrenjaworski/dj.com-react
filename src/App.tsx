import { useState, useEffect } from "react";
import journalismData from "./data/journalism.json";

type Theme = "light" | "dark" | null;
type Page = "home" | "journalism";

const SOCIAL_LINKS = [
  { name: "Instagram", url: "https://www.instagram.com/darrenjaws/" },
  { name: "Bluesky", url: "https://bsky.app/profile/darrenjaws.bsky.social" },
  { name: "Flickr", url: "https://flickr.com/photos/156906593@N08/" },
  { name: "LinkedIn", url: "https://www.linkedin.com/in/darrenjaworski" },
  { name: "Github", url: "https://github.com/darrenjaworski" },
  { name: "Strava", url: "https://www.strava.com/athletes/3266824" },
];

const RESUME_URL =
  "https://docs.google.com/document/d/19L1W3PXUyboaUWB0shDedjKwzRPsqiBw0VxsAir45EU/edit?usp=sharing";

function App() {
  const [currentPage, setCurrentPage] = useState<Page>("home");
  const [theme, setTheme] = useState<Theme>(null);

  const isSystemDark = () =>
    window.matchMedia?.("(prefers-color-scheme: dark)").matches;
  const isDarkMode = theme === "dark" || (theme === null && isSystemDark());

  useEffect(() => {
    if (theme) {
      document.documentElement.setAttribute("data-theme", theme);
    } else {
      document.documentElement.removeAttribute("data-theme");
    }
  }, [theme]);

  const handleNavClick = (page: Page, e: React.MouseEvent) => {
    e.preventDefault();
    setCurrentPage(page);
  };

  const toggleTheme = () => {
    if (theme === null) {
      setTheme(isSystemDark() ? "light" : "dark");
    } else {
      setTheme(theme === "light" ? "dark" : "light");
    }
  };

  const SocialLinks = () => (
    <>
      {SOCIAL_LINKS.map((link, index) => (
        <span key={link.name}>
          <a href={link.url}>{link.name}</a>
          {index < SOCIAL_LINKS.length - 1 ? ", " : ""}
        </span>
      ))}
    </>
  );

  const Navigation = () => (
    <nav>
      <ul>
        <li>
          <a href="/" onClick={(e) => handleNavClick("home", e)}>
            home
          </a>
        </li>
        <li>
          <a
            href="/journalism"
            onClick={(e) => handleNavClick("journalism", e)}
          >
            journalism
          </a>
        </li>
      </ul>
    </nav>
  );

  const ThemeToggle = () => (
    <button onClick={toggleTheme} aria-label="Toggle theme">
      {isDarkMode ? "‚òÄÔ∏è" : "üåô"}
    </button>
  );

  const HomePage = () => (
    <>
      <h1>home</h1>
      <div>
        <p>
          Welcome. This is my home on the web. I dramatically simplified the
          site and I hope that you like it. I'm always available for public
          comment. Please send an email to{" "}
          <a href="mailto:darrenjaworski@gmail.com">darrenjaworski@gmail.com</a>
          . (Please allow 10-15 months for response.)
        </p>

        <p>
          Check out my past work in{" "}
          <a
            href="/journalism"
            onClick={(e) => handleNavClick("journalism", e)}
          >
            journalism.
          </a>
        </p>

        <p>
          As always you can find me on <SocialLinks />
          ... (I'm sure I'm missing others. You get the point. Troll away.)
        </p>

        <p>
          For those interested in watching me wear a tie and answer questions:{" "}
          <a href={RESUME_URL}>r√©sum√©</a>.
        </p>

        <p>Now with dark theme.</p>
      </div>
    </>
  );

  const JournalismPage = () => (
    <>
      <h1>journalism</h1>
      <div>
        {journalismData.map((section) => (
          <div key={section.section}>
            <h2>{section.section}</h2>
            <ul>
              {section.articles.map((article) => (
                <li key={article.title}>
                  <a
                    href={article.link}
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    {article.title}
                  </a>
                </li>
              ))}
            </ul>
          </div>
        ))}
      </div>
    </>
  );

  return (
    <div className="main-container">
      <Navigation />
      <main>{currentPage === "home" ? <HomePage /> : <JournalismPage />}</main>
      <ThemeToggle />
    </div>
  );
}

export default App;
