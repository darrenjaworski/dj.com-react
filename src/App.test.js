import React from "react";
import renderer from "react-test-renderer";
import { shallow, mount, render } from "enzyme";

import App from "./App";

describe("app", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should be able to toggle light and dark mode", () => {});

  it("should store the theme mode in localstorage", () => {});

  it("should render with a dark background when in dark mode", () => {
    const wrapper = mount(<App />);
    wrapper.setState({ theme: "dark" });
    expect(wrapper.state("theme")).toBe("dark");
  });

  it("should render with a light background when in light mode", () => {
    const wrapper = mount(<App />);
    wrapper.setState({ theme: "light" });
    expect(wrapper.state("theme")).toBe("light");
  });

  it("should toggle between modes", () => {
    const clickWrapper = mount(<App />);
    const toggle = clickWrapper.find(".toggle");

    toggle.simulate("click");
    expect(clickWrapper.state.theme).toBe("light");

    toggle.simulate("click");
    expect(clickWrapper.state.theme).toBe("dark");
  });
});
