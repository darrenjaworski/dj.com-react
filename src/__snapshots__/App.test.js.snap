// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders without crashing 1`] = `
<div
  className="css-evlwd0"
>
  <div
    className="css-1ppe5cl"
  >
    <header>
      <nav
        className="css-1cd9ler"
      >
        <a
          aria-current="page"
          className="active"
          href="/"
          onClick={[Function]}
          style={
            Object {
              "color": "#D6D8C9",
              "fontSize": "1.25rem",
              "marginRight": "0.5rem",
              "textDecoration": "underline",
              "transition": "color 0.5s",
            }
          }
        >
          home
        </a>
        <a
          aria-current={null}
          href="/projects"
          onClick={[Function]}
          style={
            Object {
              "color": "#D6D8C9",
              "fontSize": "1.25rem",
              "marginRight": "0.5rem",
              "textDecoration": "underline",
              "transition": "color 0.5s",
            }
          }
        >
          projects
        </a>
        <a
          aria-current={null}
          href="/journalism"
          onClick={[Function]}
          style={
            Object {
              "color": "#D6D8C9",
              "fontSize": "1.25rem",
              "marginRight": "0.5rem",
              "textDecoration": "underline",
              "transition": "color 0.5s",
            }
          }
        >
          journalism
        </a>
      </nav>
    </header>
    <main>
      <h1
        className="css-1oezh8i"
      >
        home
      </h1>
      <p
        className="css-dtn5f5"
      >
        Welcome. This is my home on the web. I dramatically simplified the site and I hope that you like it. I'm always available for public comment. Please send an email to
         
        <a
          className="css-o6ojr2"
          href="mailto:darrenjaworski@gmail.com"
        >
          darrenjaworski@gmail.com
        </a>
        . 
        (Please allow 10-15 months for response.)
      </p>
      <p
        className="css-dtn5f5"
      >
        Check out my past work in
         
        <a
          href="/journalism"
          onClick={[Function]}
        >
          <a
            className="css-o6ojr2"
          >
            journalism
          </a>
        </a>
         and what I'm currently hacking away at in
         
        <a
          href="/projects"
          onClick={[Function]}
        >
          <a
            className="css-o6ojr2"
          >
            projects
          </a>
        </a>
        .
      </p>
      <p
        className="css-dtn5f5"
      >
        As always you can find me on
         
        <a
          className="css-o6ojr2"
          href="https://twitter.com/darrenjaws"
        >
          Twitter
        </a>
        ,
         
        <a
          className="css-o6ojr2"
          href="https://www.facebook.com/DarrenJJaworski"
        >
          Facebook
        </a>
        ,
         
        <a
          className="css-o6ojr2"
          href="https://www.instagram.com/darrenjaws/"
        >
          Instagram
        </a>
        ,
         
        <a
          className="css-o6ojr2"
          href="https://www.linkedin.com/in/darrenjaworski"
        >
          LinkedIn
        </a>
        , 
        <a
          className="css-o6ojr2"
          href="https://github.com/darrenjaworski"
        >
          Github
        </a>
        ,
         
        <a
          className="css-o6ojr2"
          href="http://bl.ocks.org/darrenjaworski"
        >
          bl.ocks.org
        </a>
        , 
        <a
          className="css-o6ojr2"
          href="https://medium.com/@DarrenJaws"
        >
          Medium
        </a>
        ,
         
        <a
          className="css-o6ojr2"
          href="https://www.strava.com/athletes/3266824"
        >
          Strava
        </a>
        ... (I'm sure I'm missing others. You get the point. Troll away.)
      </p>
      <p
        className="css-dtn5f5"
      >
        For those interested in watching me wear a tie and answer questions:
         
        <a
          className="css-o6ojr2"
          href="https://docs.google.com/document/d/19L1W3PXUyboaUWB0shDedjKwzRPsqiBw0VxsAir45EU/edit?usp=sharing"
        >
          résumé
        </a>
        .
      </p>
      <p
        className="css-dtn5f5"
      >
        Now with dark theme.
      </p>
    </main>
    <button
      className="toggle css-1i22e7o"
      onClick={[Function]}
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="5"
        />
        <line
          x1="12"
          x2="12"
          y1="1"
          y2="3"
        />
        <line
          x1="12"
          x2="12"
          y1="21"
          y2="23"
        />
        <line
          x1="4.22"
          x2="5.64"
          y1="4.22"
          y2="5.64"
        />
        <line
          x1="18.36"
          x2="19.78"
          y1="18.36"
          y2="19.78"
        />
        <line
          x1="1"
          x2="3"
          y1="12"
          y2="12"
        />
        <line
          x1="21"
          x2="23"
          y1="12"
          y2="12"
        />
        <line
          x1="4.22"
          x2="5.64"
          y1="19.78"
          y2="18.36"
        />
        <line
          x1="18.36"
          x2="19.78"
          y1="5.64"
          y2="4.22"
        />
      </svg>
    </button>
  </div>
</div>
`;
